src/ChunkedDecoder.php

```diff
@@ -0,0 +1,197 @@
+                $this->emit('error', array(new \Exception('Unable to identify ' . $hexValue . 'as hexadecimal number')));
+                $this->close();
+                return;
+            }
+
+            $this->chunkSize = hexdec($hexValue);
+            $this->chunkHeaderComplete = true;
+
+            $data = substr($this->buffer . $data, strlen($hexValue) + 2);
+            $this->buffer = '';
+            // Chunk header is complete
+            return $data;
+        }
+
+        $this->buffer .= $data;
+        $data = '';
+        // Chunk header isn't complete, buffer
+        return $data;
+    }
+
+    /**
+     * Extracts the chunk data and removes it from the income data string
+     *
+     * @param unknown $data - string without the hexadecimal header
+     * @return string
+     */
+    private function handleChunkData($data)
+    {
+        $chunk = substr($this->buffer . $data, 0, $this->chunkSize);
+        $this->actualChunksize = strlen($chunk);
+
+        if ($this->chunkSize == $this->actualChunksize) {
+            $data = $this->sendChunk($data, $chunk);
+        } elseif ($this->actualChunksize < $this->chunkSize) {
+            $this->buffer .= $data;
+            $data = '';
+        }
+
+        return $data;
+    }
+
+    /**
+     * Sends the chunk or ends the stream
+     *
+     * @param string $data - incomed data stream the chunk will be removed from this string
+     * @param string $chunk - chunk which will be emitted
+     * @return string - rest data string
+     */
+    private function sendChunk($data, $chunk)
+    {
+        if ($this->chunkSize == 0 && $this->isLineComplete($this->buffer . $data, $chunk, $this->chunkSize)) {
+            $this->emit('end', array());
+            return;
+        }
+
+        if (!$this->isLineComplete($this->buffer . $data, $chunk, $this->chunkSize)) {
+            $this->emit('error', array(new \Exception('Chunk doesn\'t end with new line delimiter')));
+            $this->close();
+            return;
+        }
+
+        $data = substr($this->buffer . $data, $this->chunkSize + 2);
+        $this->emit('data', array($chunk));
+
+        $this->buffer = '';
+        $this->chunkSize = 0;
+        $this->chunkHeaderComplete = false;
+
+        return $data;
+    }
+
+    /**
+     * Checks if the given chunk is ending with a "\r\n" at the start of the data string
+     *
+     * @param string $data - complete data string
+     * @param string $chunk - string which should end with "\r\n"
+     * @param unknown $length - possible length of the data chunk
+     * @return boolean
+     */
+    private function isLineComplete($data, $chunk, $length)
+    {
+        if (substr($data, 0, $length + 2) == $chunk . static::CRLF) {
+            return true;
+        }
+        return false;
+    }
+
+    /** @internal */
+    public function handleEnd()
+    {
+        if (! $this->closed) {
+            $this->emit('end');
+            $this->close();
+        }
```

Comment:
What if there's still data in the buffer when the underlying stream emits an `end` event?